### PR TITLE
SAMA5D2 - Add Flexcom SPI support

### PR DESCRIPTION
**Pull request by TimJTi**
Flexcom SPI is now working for  SAMA5D2

Created sam_flexcom_spi.c

Filled out sam_flexcom_spi.h

## Testing
Custom board with SAMA5D27C-D1G with 1 Flex SPI connected device (MAX31855) on board.

### CODE REVIEW

**Comment by TimJTi** [comment]
I welcome comments - especially style and coding standard related as I usually mess that up big time!

The DMA support is unlikely to be correct since SAMA5D2 uses XDMA not DMA and it looks like it is not done right for sam_spi.c either (and that's what this is based on).

I will revisit DMA at a later date.

File: arch/arm/src/sama5/hardware/sam_flexcom_spi.h [hunk 1]
```diff
@@ -34,10 +34,263 @@
...
+
+/* SPI FIFO Level Register */
+#define FLEXCOM_SPI_FLR_TXFL_SHIFT        (1 << 0)  /* Bits 0-5: Transmit FIFO Level */
+#define FLEXCOM_SPI_FLR_TXFL_MASK         (0x3F << FLEXCOM_SPI_FLR_TXFL_SHIFT)
+#define FLEXCOM_SPI_FLR_RXFL_SHIFT        (16 << 0)  /* Bits 16-21: Transmit FIFO Level */
+#define FLEXCOM_SPI_FLR_RXFL_MASK         (0x3F << FLEXCOM_SPI_FLR_RXFL_SHIFT)
+
+/* SPI Comparison Register */
+#define FLEXCOM_SPI_CMPR_VAL1_SHIFT       (1 << 0)  /* Bits 0-15: First Comparison Value for Received Character */
+#define FLEXCOM_SPI_CMPR_VAL1_MASK        (0xFFFF << FLEXCOM_SPI_CMPR_VAL1_SHIFT)
```
Comment (pkarashchenko) [inline arch/arm/src/sama5/hardware/sam_flexcom_spi.h:279]: please use either upper or lower case for hex constants, but not mix (I think lower is more preferred).

File: arch/arm/src/sama5/sam_flexcom_spi.h [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+ *   dev - SPI device info
+ *   devid - Identifies the (logical) device
+ *
+ * Returned Value:
+ *   Bit-encoded SPI status (see include/nuttx/spi/spi.h.
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_SAMA5_FLEXCOM0_SPI
+uint8_t sam_flexcom0_spistatus(FAR struct spi_dev_s *dev, uint32_t devid);
```
Comment (pkarashchenko) [inline arch/arm/src/sama5/sam_flexcom_spi.h:209]: remove `FAR` from this PR

File: arch/arm/src/sama5/sam_flexcom_spi.h [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+#define __FLEXCOM_SPI_CS_MASK   (1 << __FLEXCOM_SPI_CS_SHIFT)
+#  define __FLEXCOM_SPI_CS0     (0 << __FLEXCOM_SPI_CS_SHIFT)
+#  define __FLEXCOM_SPI_CS1     (1 << __FLEXCOM_SPI_CS_SHIFT)
+#define __FLEXCOM_SPI_SPI_SHIFT (1) /* Bit 1-3: Flexcom  number */
+#define __FLEXCOM_SPI_SPI_MASK  (7 << __FLEXCOM_SPI_SPI_SHIFT)
+#  define __FLEXCOM0_SPI    (0 << __FLEXCOM_SPI_SPI_SHIFT) /* FLEXCOM0 SPI */
+#  define __FLEXCOM1_SPI    (1 << __FLEXCOM_SPI_SPI_SHIFT) /* FLEXCOM1 SPI */
+#  define __FLEXCOM2_SPI    (2 << __FLEXCOM_SPI_SPI_SHIFT) /* FLEXCOM2 SPI */
+#  define __FLEXCOM3_SPI    (3 << __FLEXCOM_SPI_SPI_SHIFT) /* FLEXCOM3 SPI */
+#  define __FLEXCOM4_SPI    (4 << __FLEXCOM_SPI_SPI_SHIFT) /* FLEXCOM4 SPI */
```
Comment (pkarashchenko) [inline arch/arm/src/sama5/sam_flexcom_spi.h:54]: could you please align this with above?

File: arch/arm/src/sama5/sam_flexcom_spi.c [hunk 4]
```diff
@@ -0,0 +1,2116 @@
...
+  else if (scbr > 254)
+    {
+      scbr = 254;
+    }
+
+  scbr = (scbr + 1) & ~1;
+
+  /* Save the new scbr value */
+
+  offset = (unsigned int)g_csroffset[flex_spics->cs];
```
Comment (pkarashchenko) [inline arch/arm/src/sama5/sam_flexcom_spi.c:1164]: I do not think that cast is needed here

File: arch/arm/src/sama5/sam_flexcom_spi.c [hunk 5]
```diff
@@ -0,0 +1,2116 @@
...
+       *
+       * SPI  CPOL NCPHA
+       * MODE
+       *  0    0    1
+       *  1    0    0
+       *  2    1    1
+       *  3    1    0
+       */
+
+      offset = (unsigned int)g_csroffset[flex_spics->cs];
```
Comment (pkarashchenko) [inline arch/arm/src/sama5/sam_flexcom_spi.c:1254]: ditto

File: arch/arm/src/sama5/sam_flexcom_spi.c [hunk 6]
```diff
@@ -0,0 +1,2116 @@
...
+
+  DEBUGASSERT(nbits == 8); /* Temporary -- FIX ME */
+
+  /* Has the number of bits changed? */
+
+  if (nbits != flex_spics->nbits)
+    {
+      /* Yes... Set number of bits appropriately */
+
+      offset  = (unsigned int)g_csroffset[flex_spics->cs];
```
Comment (pkarashchenko) [inline arch/arm/src/sama5/sam_flexcom_spi.c:1327]: ditto

File: arch/arm/src/sama5/sam_flexcom_spi.c [hunk 7]
```diff
@@ -0,0 +1,2116 @@
...
+#endif
+
+      flex_spi_dumpregs(flex_spi, "After initialization");
+    }
+
+  /* Set to mode=0 and nbits=8 and impossible frequency. The SPI will only
+   * be reconfigured if there is a change.
+   */
+
+  offset = (unsigned int)g_csroffset[flex_csno];
```
Comment (pkarashchenko) [inline arch/arm/src/sama5/sam_flexcom_spi.c:2102]: ditto

File: arch/arm/src/sama5/sam_flexcom_spi.c [hunk 8]
```diff
@@ -0,0 +1,2116 @@
...
+      return &g_flexcom2dev;
+#endif
+#ifdef CONFIG_SAMA5_FLEXCOM3_SPI
+  if (flex_spics->flex_spino == 3)
+      return &g_flexcom3dev;
+#endif
+#ifdef CONFIG_SAMA5_FLEXCOM4_SPI
+  if (flex_spics->flex_spino == 4)
+      return &g_flexcom4dev;
+#endif
```
Comment (pkarashchenko) [inline arch/arm/src/sama5/sam_flexcom_spi.c:687]: maybe better convert to switch-case?

File: arch/arm/src/sama5/sam_flexcom_spi.c [hunk 9]
```diff
@@ -0,0 +1,2116 @@
...
+ *   flex_spics - Device-specific state data
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+static inline uint32_t flex_spi_cs2pcs(struct sam_flex_spics_s *flex_spics)
+{
+  return ((uint32_t)1 << (flex_spics->cs)) - 1;
```
Comment (pkarashchenko) [inline arch/arm/src/sama5/sam_flexcom_spi.c:752]:
```suggestion
  return ((uint32_t)1 << flex_spics->cs) - 1;
```

File: arch/arm/src/sama5/sam_flexcom_spi.c [hunk 10]
```diff
@@ -0,0 +1,2116 @@
...
+ *   frequency - The SPI frequency requested
+ *
+ * Returned Value:
+ *   Returns the actual frequency selected
+ *
+ ****************************************************************************/
+
+static void flex_spi_select(struct spi_dev_s *dev, uint32_t devid,
+                       bool selected)
+{
```
Comment (pkarashchenko) [inline arch/arm/src/sama5/sam_flexcom_spi.c:1065]:
```suggestion
static void flex_spi_select(struct spi_dev_s *dev, uint32_t devid,
                            bool selected)
{
```

File: arch/arm/src/sama5/sam_flexcom_spi.c [hunk 11]
```diff
@@ -0,0 +1,2116 @@
...
+        case SPIDEV_MODE2: /* CPOL=1; NCPHA=1 */
+          regval |= (FLEXCOM_SPI_CSR_CPOL | FLEXCOM_SPI_CSR_NCPHA);
+          break;
+
+        case SPIDEV_MODE3: /* CPOL=1; NCPHA=0 */
+          regval |= FLEXCOM_SPI_CSR_CPOL;
+          break;
+
+        default:
+          DEBUGASSERT(FALSE);
```
Comment (pkarashchenko) [inline arch/arm/src/sama5/sam_flexcom_spi.c:1276]:
```suggestion
          DEBUGASSERT(false);
```

File: arch/arm/src/sama5/sam_flexcom_spi.c [hunk 12]
```diff
@@ -0,0 +1,2116 @@
...
+      if (rxptr)
+        {
+          *rxptr++ = (uint8_t)data;
+        }
+    }
+}
+
+#ifdef CONFIG_SAMA5_SPI_DMA
+static void flex_spi_exchange(struct spi_dev_s *dev, const void *txbuffer,
+                         void *rxbuffer, size_t nwords)
```
Comment (pkarashchenko) [inline arch/arm/src/sama5/sam_flexcom_spi.c:1519]:
```suggestion
static void flex_spi_exchange(struct spi_dev_s *dev, const void *txbuffer,
                              void *rxbuffer, size_t nwords)
```

File: arch/arm/src/sama5/sam_flexcom_spi.c [hunk 13]
```diff
@@ -0,0 +1,2116 @@
...
+ *            uint16_t's
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+#ifndef CONFIG_SPI_EXCHANGE
+static void flex_spi_sndblock(struct spi_dev_s *dev, const void *buffer,
+                         size_t nwords)
```
Comment (pkarashchenko) [inline arch/arm/src/sama5/sam_flexcom_spi.c:1766]:
```suggestion
static void flex_spi_sndblock(struct spi_dev_s *dev, const void *buffer,
                              size_t nwords)
```

File: arch/arm/src/sama5/sam_flexcom_spi.c [hunk 14]
```diff
@@ -0,0 +1,2116 @@
...
+ *            packed into uint16_t's
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+#ifndef CONFIG_SPI_EXCHANGE
+static void flex_spi_recvblock(struct spi_dev_s *dev, void *buffer,
+                          size_t nwords)
```
Comment (pkarashchenko) [inline arch/arm/src/sama5/sam_flexcom_spi.c:1796]:
```suggestion
static void flex_spi_recvblock(struct spi_dev_s *dev, void *buffer,
                               size_t nwords)
```

File: arch/arm/src/sama5/sam_flexcom_spi.c [hunk 15]
```diff
@@ -0,0 +1,2116 @@
...
+        flex_spics->candma = SAMA5_FLEXCOM2_SPI_DMA;
+        break;
+      case 3:
+        flex_spics->candma = SAMA5_FLEXCOM3_SPI_DMA;
+        break;
+      case 4:
+        flex_spics->candma = SAMA5_FLEXCOM4_SPI_DMA;
+        break;
+      default:
+        DEBUGASSERT(FALSE);
```
Comment (pkarashchenko) [inline arch/arm/src/sama5/sam_flexcom_spi.c:1890]:
```suggestion
        DEBUGASSERT(false);
```

File: arch/arm/src/sama5/sam_flexcom_spi.c [hunk 16]
```diff
@@ -0,0 +1,2116 @@
...
+           */
+        }
+#endif
+
+#ifdef CONFIG_SAMA5_FLEXCOM2_SPI
+      if (flex_spino == 2)
+        {
+          sam_flexcom2_enableclk();
+
+          /* Configure multiplexed pins as connected on the board,
```
Comment (pkarashchenko) [inline arch/arm/src/sama5/sam_flexcom_spi.c:2015]: just an option: switch-case

File: arch/arm/src/sama5/hardware/sam_flexcom_spi.h (hunk 1 above)
Comment (TimJTi) [reply arch/arm/src/sama5/hardware/sam_flexcom_spi.h:279]: Ah, yes - sorry. I actually created this file 18 months ago before I'd paid any attention to the style guide and coding standards. I'll go through the files and check for all of these.

File: arch/arm/src/sama5/sam_flexcom_spi.c (hunk 8 above)
Comment (TimJTi) [reply arch/arm/src/sama5/sam_flexcom_spi.c:687]: I followed the method in sam_spi.c (which is just for SPI0 and SPI1), and changed/extended it. We need to make sure the test isn't done for any interfaces not actually in use, but I'll check to see if we would ever get to this code if not and revise as appropriate.

File: arch/arm/src/sama5/sam_flexcom_spi.h (hunk 2 above)
Comment (TimJTi) [reply arch/arm/src/sama5/sam_flexcom_spi.h:209]: ok

File: arch/arm/src/sama5/sam_flexcom_spi.c (hunk 9 above)
Comment (TimJTi) [reply arch/arm/src/sama5/sam_flexcom_spi.c:752]: ok

File: arch/arm/src/sama5/sam_flexcom_spi.c (hunk 10 above)
Comment (TimJTi) [reply arch/arm/src/sama5/sam_flexcom_spi.c:1065]: ok

File: arch/arm/src/sama5/sam_flexcom_spi.c (hunk 11 above)
Comment (TimJTi) [reply arch/arm/src/sama5/sam_flexcom_spi.c:1276]: ok

File: arch/arm/src/sama5/sam_flexcom_spi.c (hunk 12 above)
Comment (TimJTi) [reply arch/arm/src/sama5/sam_flexcom_spi.c:1519]: missed that - ok

File: arch/arm/src/sama5/sam_flexcom_spi.c (hunk 13 above)
Comment (TimJTi) [reply arch/arm/src/sama5/sam_flexcom_spi.c:1766]: ok

File: arch/arm/src/sama5/sam_flexcom_spi.c (hunk 14 above)
Comment (TimJTi) [reply arch/arm/src/sama5/sam_flexcom_spi.c:1796]: ok

File: arch/arm/src/sama5/sam_flexcom_spi.c (hunk 15 above)
Comment (TimJTi) [reply arch/arm/src/sama5/sam_flexcom_spi.c:1890]: ok

File: arch/arm/src/sama5/sam_flexcom_spi.c (hunk 16 above)
Comment (TimJTi) [reply arch/arm/src/sama5/sam_flexcom_spi.c:2015]: as above, it's as per sam_spi.c so will double check and revise as appropriate

File: arch/arm/src/sama5/sam_flexcom_spi.c (hunk 4 above)
Comment (TimJTi) [reply arch/arm/src/sama5/sam_flexcom_spi.c:1164]: I see now that sam_spi.c has g_csroffset as unit8_t as the offsets are within the range of a byte, but for flexcom they are not so I changed it to uint16_t here. "offset" is an unsigned int which is 32 bits on this processor so I think the cast may technically be unnecessary as no information is likely to be lost without the case, but isn't it good practice to do this?

File: arch/arm/src/sama5/sam_flexcom_spi.c (hunk 8 above)
Comment (TimJTi) [reply arch/arm/src/sama5/sam_flexcom_spi.c:687]: OK - I've misunderstoof what sam_spi.c code was doing here so this is actually wrong. I'll demote this PR back to being a draft while I resolve it.

A good but unintentional find - thanks :)

Comment (TimJTi) [reply arch/arm/src/sama5/sam_flexcom_spi.c:687]: no, I didn't misunderstand but the comments in sam_spi.c were misleading.

@pkarashchenko - I'm not sure a switch method is any neater, unless I misunderstood?


```suggestion
  switch (flex_spics->spino)
    {
#ifdef CONFIG_SAMA5_FLEXCOM0_SPI
     case 0:
      return &g_flexcom0dev;
      break;
 #endif
 #ifdef CONFIG_SAMA5_FLEXCOM1_SPI
      case 1:
        return &g_flexcom1dev;
        break;
#endif
etc
```

File: arch/arm/src/sama5/sam_flexcom_spi.c (hunk 16 above)
Comment (TimJTi) [reply arch/arm/src/sama5/sam_flexcom_spi.c:2015]: @pkarashchenko same comment as earlier - is it actually neater as a switch?

Comment (pkarashchenko) [reply arch/arm/src/sama5/sam_flexcom_spi.c:2015]: I was thinking about something like
```
#if defined(CONFIG_SAMA5_FLEXCOM0_SPI) || \
    defined(CONFIG_SAMA5_FLEXCOM1_SPI) || \
    defined(CONFIG_SAMA5_FLEXCOM2_SPI) || \
    defined(CONFIG_SAMA5_FLEXCOM3_SPI) || \
    defined(CONFIG_SAMA5_FLEXCOM4_SPI)
#  define SAMA5_FLEXCOM_SPINO
#endif
...
#ifdef SAMA5_FLEXCOM_SPINO
  uint8_t flex_spino;               /* SPI controller number (0 or 1) */
#endif
...
#ifdef SAMA5_FLEXCOM_SPINO
  switch (flex_spino)
    {
#ifdef CONFIG_SAMA5_FLEXCOM0_SPI
      case 0:
        sam_flexcom0_enableclk();
        break;
#endif
...
    }
#endif
```

Comment (TimJTi) [reply arch/arm/src/sama5/sam_flexcom_spi.c:2015]: Aha! Much better idea and is then consistent with the flexcom usart code. Thanks :)

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**